### ISSUE TITLE
Add API to obtain current injection point from Bean#create

### ISSUE THREAD

**Issue by arjantijms** [jakartaee/cdi#637]
As per the existing issue https://issues.redhat.com/browse/CDI-610

There's currently not a clear way on how to obtain the current injection point (if any) from `Bean<T>#create`.
A previously somewhat accepted way (though not specified) was:

```java
  Bean<? extends Object> bean = beanManager.resolve(beanManager.getBeans(InjectionPoint.class));
  InjectionPoint injectionPoint = (InjectionPoint) beanManager.getReference(bean, InjectionPoint.class, creationalContext);
```

This however broke in some version of Weld.

Since getting the injection point is an often used feature in producers, I'd like to propose to introduce an easy API for this, so `Bean<T>` implementations can use this just as easily. E.g. something like: `BeanManager#getCurrentInjectionPoint()`.

Also see: http://cdi-development-mailing-list.1064426.n5.nabble.com/Getting-injection-point-from-Bean-create-td5710505i20.html

**Comment by arjantijms** [comment]
p.s. a workaround is now:

```java
@Dependent
public class InjectionPointGenerator {

    // TODO: this is a workaround originally for older OWB versions, but while OWB is fixed, newer Weld
    // versions are now broken. It seems this needs to be fixed in CDI 2.1, or perhaps CDI 3.
    // See https://issues.jboss.org/browse/CDI-610

    @Inject
    private InjectionPoint injectionPoint;

    public InjectionPoint getInjectionPoint() {
        return injectionPoint;
    }

}
```

```java
    public static InjectionPoint getCurrentInjectionPoint(BeanManager beanManager, CreationalContext<?> creationalContext) {
        Bean<InjectionPointGenerator> bean = resolve(beanManager, InjectionPointGenerator.class);

        return bean != null
                ? (InjectionPoint) beanManager.getInjectableReference(bean.getInjectionPoints().iterator().next(), creationalContext)
                : null;
    }

    @SuppressWarnings("unchecked")
    public static <T> Bean<T> resolve(BeanManager beanManager, Class<T> beanClass, Annotation... qualifiers) {
        Set<Bean<?>> beans = beanManager.getBeans(beanClass, qualifiers);

        for (Bean<?> bean : beans) {
            if (bean.getBeanClass() == beanClass) {
                return (Bean<T>) beanManager.resolve(Collections.<Bean<?>>singleton(bean));
            }
        }

        Bean<T> bean = (Bean<T>) beanManager.resolve(beans);

        if (bean == null && beanClass.getSuperclass() != Object.class) {
            return (Bean<T>) resolve(beanManager, beanClass.getSuperclass(), qualifiers);
        } else {
            return bean;
        }
    }
```

**Comment by Ladicek** [comment]
I agree with @mkouba's comments in the original issue -- the existing CDI API for programmatically looking up `InjectionPoint` is sufficient. The spec should clarify that this is possible, but only if the "currently injected" bean is `@Dependent` (which is probably gonna be interesting to put into text :-) ).

> This however broke in some version of Weld.

I'd say that's probably a bug, but I'll let @manovotn comment on that :-)

**Comment by manovotn** [comment]
> > This however broke in some version of Weld.
> 
> I'd say that's probably a bug, but I'll let @manovotn comment on that :-)

That sounds like a bug and I am not aware of any issue tracking that.
The linked CDI issue even shows Weld tests that assert this very thing.
@arjantijms could you create a WELD issue with a test reproducing it for the case where this doesn't work?
I'd be happy to look into it but I need to know what is the scenario where this fails.

As for this issue, I am also of the opinion that you should be able to use programmatic lookup to obtain `InjectionPoint`.

**Comment by arjantijms** [comment]
I just discovered this one is still open after all this time. I don't know exactly which Weld version it was back then, but it's not an issue with Weld anymore at least.

Just did a quick test on latest Weld (7.0.0.Alpha1) and in the following the right  `InjectionPoint` is obtained

```
public void foo(@Observes AfterBeanDiscovery afterBeanDiscovery, BeanManager beanManager) {

    afterBeanDiscovery.addBean()
                    .scope(Dependent.class)
                    .beanClass(...)
                    .types(...)
                    .qualifiers(...)
                    .id(...)
                    .createWith(creationalContext -> {

                        InjectionPoint injectionPoint = beanManager.createInstance().select(InjectionPoint.class).get();
```

Was it ever added to the spec with a TCK test that this is the supported way to do it?

**Comment by manovotn** [comment]
> Was it ever added to the spec with a TCK test that this is the supported way to do it?

A test alone would not be enough, we would need to cover this with a spec text and then provide a TCK.

BTW, is there a reason you are doing this with a BM from the method parameters instead of using `produceWith(Function<Instance<Object>, U> callback);` where you have access to the `Instance` directly?
That's what I would perceive as a natural way to obtain an IP related to that given bean (though both work as far as I know).

> I just discovered this one is still open after all this time. 

Yea, I wonder if we should (after 5.0) do a sweep of existing issues, asking for feedback where needed and closing those that get no answer or no clear way forward. Something we might want to discuss on the next CDI meeting @Ladicek, @Azquelt

**Comment by Ladicek** [comment]
I'm fairly sure we have a TCK test for doing this at least for Build Compatible Extensions. I don't know if there's one for Portable Extensions, but it feels likely to me.

IMHO, this doesn't need new spec text or new TCKs (although if a test for PEs is missing, one should definitely be added).

**Comment by manovotn** [comment]
@arjantijms with the TCK test in place, do you agree we can close this issue?

**Comment by arjantijms** [comment]
> > Was it ever added to the spec with a TCK test that this is the supported way to do it?
> 
> A test alone would not be enough, we would need to cover this with a spec text and then provide a TCK.
> 
> BTW, is there a reason you are doing this with a BM from the method parameters instead of using `produceWith(Function<Instance<Object>, U> callback);` where you have access to the `Instance` directly? That's what I would perceive as a natural way to obtain an IP related to that given bean (though both work as far as I know).

There's no specific reason really, other than that I quickly tested it in a piece of code that already had that bean manager, and that the [Weld test ](https://github.com/weld/core/blob/master/tests-arquillian/src/test/java/org/jboss/weld/tests/injectionPoint/custom/JuicyBarBean.java#L41) you mentioned did something similar.

**Comment by arjantijms** [comment]
> with the TCK test in place, do you agree we can close this issue?

Yes, it probably can. 

Maybe we want to explicitly test that the following also works? I don't have strong feelings about it, but for completeness may be worth considering:

https://github.com/weld/core/blob/master/tests-arquillian/src/test/java/org/jboss/weld/tests/injectionPoint/custom/JuicyBarBean.java#L41

**Comment by manovotn** [comment]
> > with the TCK test in place, do you agree we can close this issue?
> 
> Yes, it probably can.
> 
> Maybe we want to explicitly test that the following also works? I don't have strong feelings about it, but for completeness may be worth considering:
> 
> https://github.com/weld/core/blob/master/tests-arquillian/src/test/java/org/jboss/weld/tests/injectionPoint/custom/JuicyBarBean.java#L41

I wouldn't mind that, PRs are welcome :)
